File: go/src/guengel.ch/statistics/statistics.go
package statistics

import (
	"fmt"
	"math"
	"os"
	"time"
)

type statisticsEntry struct {
	min        int64
	max        int64
	dataPoints []int64
}

type Statistics struct {
	statistics map[string]statisticsEntry
}

func minInt64(a, b int64) int64 {
	if a < b {
		return a
	}
	return b
}

func maxInt64(a, b int64) int64 {
	if a > b {
		return a
	}
	return b
}

func newStatisticsEntry(label string) *statisticsEntry {
	return &statisticsEntry{
		math.MaxInt64,
		math.MinInt64,
		make([]int64, 0),
	}
}

func (se *statisticsEntry) addDataPoint(value int64) {
	se.dataPoints = append(se.dataPoints, value)
	se.max = maxInt64(se.max, value)
	se.min = minInt64(se.min, value)
}

func (se *statisticsEntry) average() int64 {
	var sum int64
	for _, v := range se.dataPoints {
		sum += v
	}

	return int64(math.Round(float64(sum) / float64(len(se.dataPoints))))
}

func Start() *Statistics {
	return &Statistics{make(map[string]statisticsEntry)}
}

func (s *Statistics) Record(label string, start, end time.Time) {
	value := end.Sub(start).Nanoseconds()

	entry, exists := s.statistics[label]
	if !exists {
		stats := *newStatisticsEntry(label)
		stats.addDataPoint(value)
		s.statistics[label] = stats
	} else {
		entry.addDataPoint(value)
		s.statistics[label] = entry
	}
}

func (s *Statistics) DumpToFile(filename string) {
	file, err := os.Create(filename)
	if err != nil {
		panic(err)
	}
	defer file.Close()

	for k, v := range s.statistics {
		stat := fmt.Sprintf("go-array;%s;%v;%v;%v\n", k, v.min, v.average(), v.max)
		file.WriteString(stat)
	}
}

File: kotlin/src/main/kotlin/support/statistics.kt
package support

import java.io.BufferedWriter
import java.io.FileWriter
import kotlin.math.round

private class StatisticsEntry {
    private val dataPoints = ArrayList<Long>()
    var max = Long.MIN_VALUE
        private set(value) {
            field = kotlin.math.max(field, value)
        }
    var min = Long.MAX_VALUE
        private set (value) {
            field = kotlin.math.min(field, value)
        }

    fun addData(value: Long) {
        max = value
        min = value
        dataPoints.add(value)
    }

    fun average(): Long = round(dataPoints.average()).toLong()
}

class Statistics {
    private val statisticsMap = HashMap<String, StatisticsEntry>()

    fun addData(label: String, start: Long, end: Long) {
        val entry = statisticsMap.getOrPut(label) { StatisticsEntry() }
        entry.addData(end - start)
    }

    fun dumpStatisticsToFile(filename: String) {
        BufferedWriter(FileWriter(filename)).use {
            for ((k, v) in statisticsMap) {
                it.write("${PREFIX};$k;${v.min};${v.average()};${v.max}\n")
            }
        }
    }
}